src/onnxruntime.cc

```diff
@@ -1453,17 +1458,69 @@ ModelInstanceState::ProcessRequests(
           &input_names, &cuda_copy));
 
   if (!all_response_failed) {
+    // Set preferred memory type and id. This will be used while querying
+    // memory type to be used for output buffer.
+    TRITONSERVER_MemoryType preferred_memory_type = TRITONSERVER_MEMORY_CPU;
+    int64_t preferred_memory_type_id = 0;
+    if (Kind() == TRITONSERVER_INSTANCEGROUPKIND_GPU) {
+      preferred_memory_type = TRITONSERVER_MEMORY_GPU;
+      preferred_memory_type_id = DeviceId();
+    }
+
     // Request to retrieve all model outputs. 'output_names' and
     // 'output_tensors_' are parallel vectors and so must be kept in
     // sync. [TODO] should collect only the outputs needed by some
     // request.
     for (auto& output_name : StateForModel()->ModelOutputs()) {
       output_tensors_.emplace_back(nullptr);
 
+      TRITONSERVER_MemoryType memory_type = preferred_memory_type;
+      int64_t memory_type_id = preferred_memory_type_id;
+      if (output_device_info_.size() !=
+          StateForModel()->ModelOutputs().size()) {
+        // Query the memory type of destination output buffer. Bind the output
+        // to this destination memory type. The destination memory type
+        // for an output for all requests should be same. So use any request for
+        // this query.
+        auto err = TRITONBACKEND_RequestOutputBufferProperties(
+            requests[0], output_name.first.c_str(), /*byte_size*/ nullptr,
+            &memory_type, &memory_type_id);
+
+        if (err != nullptr) {
+          LOG_MESSAGE(
+              TRITONSERVER_LOG_WARN,
```

Comment:
Make this verbose, otherwise you will get a warning for every request to the model.